File: py/expression_add_operators.py
# https://leetcode.com/problems/expression-add-operators/

from turtle import back
from typing import List


class Solution:
    def addOperators(self, num: str, target: int) -> List[str]:
        res = []

        def backtrack(pos, path, result_so_far, prev_value):
            if pos == len(num):
                if result_so_far == target:
                    res.append(path)
                return
            for i in range(pos, len(num)):
                if i > pos and num[pos] == "0":
                    break  # skip leading zero number. e.g. 1,02
                cur = int(num[pos : i + 1])
                if pos == 0:
                    backtrack(i + 1, path + str(cur), cur, cur)
                else:
                    backtrack(i + 1, path + "+" + str(cur), result_so_far + cur, cur)
                    backtrack(i + 1, path + "-" + str(cur), result_so_far - cur, -cur)
                    backtrack(
                        i + 1,
                        path + "*" + str(cur),
                        result_so_far - prev_value + prev_value * cur,
                        prev_value * cur,
                    )  # going backward and re-eval. e.g. 2+3*2: result_so_far = 5 (2+3); prev_value = 3; eval = 5 - 3 + 3 * 2

        backtrack(0, "", 0, 0)
        return res


s = Solution()
num, target = "105", 5
print(s.addOperators(num, target))

File: src/BackTrack/ExpressionAddOperators.java
package BackTrack;

import java.util.ArrayList;
import java.util.List;

// https://leetcode.com/problems/expression-add-operators/
public class ExpressionAddOperators {
    public List<String> addOperators(String num, int target) {
        List<String> rst = new ArrayList<>();
        if (num == null || num.length() == 0) return rst;
        helper(rst, "", num, target, 0, 0, 0);
        return rst;
    }

    public void helper(List<String> rst, String path, String num, int target, int pos, long previousEval, long previousValue) {
        if (pos == num.length()) {
            if (target == previousEval)
                rst.add(path);
            return;
        }
        for (int i = pos; i < num.length(); i++) {
            if (i != pos && num.charAt(pos) == '0') break; // don't accept leading zeros
            long cur = Long.parseLong(num.substring(pos, i + 1)); // 2,3,2/2,32/23,2/232
            if (pos == 0) {
                helper(rst, path + cur, num, target, i + 1, cur, cur);
            } else {
                helper(rst, path + "+" + cur, num, target, i + 1, previousEval + cur, cur);

                helper(rst, path + "-" + cur, num, target, i + 1, previousEval - cur, -cur);

                // going backward and re-eval. e.g. 2+3*2: previousEval = 5 (2+3); previousValue = 3; current eval = 5 - 3 + 3 * 2
                helper(rst, path + "*" + cur, num, target, i + 1, previousEval - previousValue + previousValue * cur, previousValue * cur);
            }
        }
    }

    public static void main(String[] args) {
        ExpressionAddOperators e = new ExpressionAddOperators();
        String input = "232";
        List<String> res = e.addOperators(input, 8);
        res.forEach(System.out::println);
    }
}
